kroxylicious/src/main/java/io/kroxylicious/proxy/internal/SaslDecodePredicate.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.internal;
+
+import org.apache.kafka.common.protocol.ApiKeys;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.kroxylicious.proxy.internal.codec.DecodePredicate;
+
+class SaslDecodePredicate implements DecodePredicate {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(SaslDecodePredicate.class);
+
+    private final boolean handleSasl;
+    private DecodePredicate delegate = null;
+
+    public SaslDecodePredicate(boolean handleSasl) {
+        this.handleSasl = handleSasl;
+    }
+
+    public void setDelegate(DecodePredicate delegate) {
```

Comment:
Something about having a delegate bothers me, but I'm struggling to express it. 

Why is the `SaslDecodePredicate` special? Why isn't it just one off the set of predicates to be invoked? 

I think its telling me that the abstraction isn't quite right that we need this this specialisation, or it this impl has too many responsibilities.

Reply:
I agree that it's ugly, but I could see a nicer way of handling the problem.

The problem is that:

1. We want the proxy to avoid deserializing requests and responses "when it doesn't have to"
    * So when there isn't a filter which is interested in that request/response API, or API version
    * And when the proxy infra itself doesn't need to
2. With this SASL offload stuff the proxy itself is now (when configured) interested in the SASL APIs.
3. But it doesn't know, until it's got to the invoking the `NetFilter` impl and that having called back on the `NetFilterContext`, what protocol filters are to be used. 
4. But it's the `KafkaDecodeFilter` which needs to know about decodability, and that sits in front of the `KafkaProxyFrontendHandler`, so there's a cyclic dependency.
5. It's easier to use this delegation pattern than it is to try to reconfigure the predicate on the `KafkaDecodeFilter`.

I'm happy to consider an alternative approach if you can describe one. But failing that I'll add a comment along the above lines, so it's at least explained.

Reply:
Let's stick to a comment for now. 

It still sort of feels like this predicate should just be the first in the list and the rest can be added/removed later but I haven't thought deeply enough about it yet. It doesn't feel like a reason not to merge the current change.